framework/audio/midi/internal/sequencer.cpp

```diff
@@ -58,6 +49,53 @@ void Sequencer::init(float samplerate, float gain)
     });
 }
 
+void Sequencer::loadMIDI(const std::shared_ptr<MidiStream>& stream)
+{
+    m_midiStream = stream;
+    m_streamState = StreamState();
+
+    m_midiData = stream->initData;
+
+    m_midiStream->stream.onReceive(this, [this](const MidiData& data) { onDataReceived(data); });
+    m_midiStream->stream.onClose(this, [this]() { onStreamClosed(); });
+
+    requestData(0);
+
+    buildTempoMap();
+    synth()->loadSF(m_midiData.programs(), "", [this](uint16_t percent) {
+        LOGI() << "sf loading: " << percent;
+    });
+}
+
+void Sequencer::requestData(uint32_t tick)
+{
+    LOGI() << "requestData: " << tick;
+    if (m_streamState.closed) {
+        LOGE() << "stream closed";
+        m_streamState.requested = false;
+        return;
+    }
+    m_streamState.requested = true;
+    m_midiStream->request.send(tick);
+}
+
+void Sequencer::onDataReceived(const MidiData& data)
+{
+    LOGI() << "onDataReceived: " << data.tracks.front().channels.front().events.front().tick;
+    //! TODO implement merge
+    m_midiData = data;
+    m_streamState.requested = false;
+
+    uint32_t curTick = ticks(m_curMsec);
+    doSeekTracks(curTick, m_midiData.tracks);
+}
+
+void Sequencer::onStreamClosed()
+{
+    m_streamState.requested = false;

```

Comment:
Isn't it better to make it to be an enum?

enum StreamState {
    OPENED,
    CLOSED
}

Or event create the private method "isStreamOpened()" because the values seem to be mutually exclusive

Reply:
Here the meaning is different, it means that when the stream is closed, we will not wait for the data to be received, so we reset this flag.

See https://github.com/musescore/MuseScore/pull/6354#discussion_r458874409